### ISSUE TITLE
AKFrequencyTracker hogs cpu, 45-50% pegged.

### ISSUE THREAD

**Issue by ghubaddole** [AudioKit/AudioKit#1753]
I have a fairly complex audio graph of AKplayers, mixers, boosters, compressors, limiters, etc.
I set them up and then call Audiokit.start(). 
I have controls to start / stop instruments.
Even when no sound is playing, the CPU is at ~44%. When sound is playing, increases to about 50-55%.
Only when I call Audiokit.stop() does the thread go away and CPU goes to effectively zero (1-2%.)

Why is audiokit hogging cpu when no sound is played? I can understand a small amount of constant load but 44% seems excessive.
This causes my device to heat up, my battery life to go to hell.

Another potentially related issue may be that when the device is warm to hot, the audio stutters a lot... probably because of the AURemoteIO thread getting throttled?

What am I missing?
-g

**Comment by ghubaddole** [comment]
It was the frequency tracker deep within this stack trace ... :-(
I wonder why it's taking 44% cpu though.

**Comment by aure** [comment]
I'll take a look at the frequency tracker. Its probably not the most efficient code but I didn't think it was so bad before, perhaps something got borked.

**Comment by sieren** [comment]
@ghubaddole any specific function call in that stacktrace? Like was it `sp_ptrack_compute` ?

**Comment by ghubaddole** [comment]
I am calling frequency tracking every 20 milliseconds. Also, it seems that calling it every 5 milliseconds doesn't raise the CPU consumption by much! 
All this time I was expecting my code to be heavy on the CPU - lot of computation, but my code is accounting for about 5% cpu!! The rest was all the AKFrequencyTracker! I'm attaching a very large and very confusing stacktrace screenshot (with app name redacted). All i'm doing is start the frequency tracker and sample it every 0.02 seconds! :-)
The bottom of the heaviest stack trace is also attached. There's also AKStringResonatorDSP at the bottom of that as well -- but I've not started the string resonators at all! No sound is playing when I capture the stack trace -- when sound is playing the CPU goes up by about 5-10% max.

Please help me make some sense of this stacktrace!

![Screen Shot 2019-04-19 at 7 40 26 AM](https://user-images.githubusercontent.com/46110201/56401647-8f850300-6277-11e9-958b-0402a0d31859.png)

![Screen Shot 2019-04-19 at 7 40 15 AM](https://user-images.githubusercontent.com/46110201/56401621-754b2500-6277-11e9-9805-fa7a9e0d7316.png)

**Comment by ghubaddole** [comment]
I'll really appreciate some help with this. If this can be fixed, a lot of issues will go away with my app, to which this is critical functionality.
Thank you in advance, Aurelius and team!

**Comment by ghubaddole** [comment]
I now shut all the resonators and only start them when I need them and now the stack trace is pretty clear and points to the freq tracker.

@sieren - sp_ptrack_compute. bingo!

![Screen Shot 2019-04-19 at 7 23 36 PM](https://user-images.githubusercontent.com/46110201/56427260-d7884200-62d8-11e9-918e-4b2d787e17fe.png)
![Screen Shot 2019-04-19 at 7 24 18 PM (2)](https://user-images.githubusercontent.com/46110201/56427295-f8509780-62d8-11e9-8dc4-1b51ce9cd008.png)

**Comment by ghubaddole** [comment]
Also found an old thread - https://groups.google.com/forum/#!msg/audiokit/J6vculPKZt0/YgJDygWEAAAJ

**Comment by sieren** [comment]
@ghubaddole the frequency tracker is based on code here: https://github.com/PaulBatchelor/Soundpipe

Unfortunately none of it uses SIMD or NEON optimizations (or abstraction layers like vDSP) because the main goal was portability I suppose.
It might be worth asking @PaulBatchelor if there's anything you could do to optimise it.

@aure Do we have any rules when it comes to the Soundpipe Library? I was thinking of optimizing some of that code using Apple's vDSP library, which obv we couldnt merge upstream because it'd make the code platform specific.

**Comment by aure** [comment]
We have our own copy of Soundpipe embedded into AudioKit, so we can do stuff like this.  I wouldn't mind making a copy of one of Paul's things, like sp_ptrack, and calling sp_ptrack_vdsp or sp_ptrack_apple.

**Comment by ghubaddole** [comment]
That would be truly awesome!!

**Comment by ghubaddole** [comment]
@sieren, thank you for looking into optimizing this! Much appreciated!

**Comment by sieren** [comment]
@ghubaddole no worries, but it may take a while before I have the time to sink my teeth into this.

**Comment by ghubaddole** [comment]
@sieren Thanks!! No pressure, but if you can make a dent in this in the next couple of weeks, It'll really help - my app release is currently gated on this one issue :-(

**Comment by ghubaddole** [comment]
@sieren, sorry to bother you but did you get a chance to look into this?

**Comment by mfk-ableton** [comment]
Hey, sorry if I wasnt clear. By a "while" I meant its probably going to take a few weeks before I have the time. I'm currently traveling and have literally zero time to look into this.

**Comment by sieren** [comment]
Hey, sorry if I wasnt clear. By a "while" I meant its probably going to take a few weeks before I have the time. I'm currently traveling and have literally zero time to look into this.

**Comment by PaulBatchelor** [comment]
@ghubaddole If you really need this done quickly, you may benefit from studying the code yourself?

The code for the ptrack algorithm is contained inside of a single C file and a header file (which gets amalgamated into a single header "soundpipe.h" via command line one-liner). I'm posting them here in case anyone wants to take a look at it.

Here they are in the Soundpipe master branch:

https://github.com/PaulBatchelor/Soundpipe/blob/master/modules/ptrack.c
https://github.com/PaulBatchelor/Soundpipe/blob/master/h/ptrack.h

Here is some sample C code using it:

https://github.com/PaulBatchelor/Soundpipe/blob/master/examples/ex_ptrack.c

Here is ptrack in AudioKit:

https://github.com/AudioKit/AudioKit/blob/master/AudioKit/Core/Soundpipe/modules/ptrack.c

I imagine if you are building AudioKit from source, you could use a drop-in replacement of ptrack, assuming the external inputs/outputs are the same.

As you can see in the top of the C source file, the original creator of the algorithm is Miller Puckette (probably as a PD object somewhere?), and was ported as Csound Opcode by Victor Lazzarini. Those are the two people you'd want to contact to really get insight into what is going on, and places to optimize.

@sieren is a very talented developer who does this stuff professionally. I have no doubt they will do amazing things to optimize this particular algorithm in AK if/when they get the time. Sometimes we all need to be reminded again from time to time that the contributors to AK do so out of the goodness of their heart, with no compensation beyond a pat on the back. This particular kind of optimization does take a non-trivial amount of time and expertise to do, and people get paid very well in the industry to do tasks like this. It would be awesome if it were to get done, but there is ZERO obligation for it to happen.

**Comment by ghubaddole** [comment]
I understand that industry experts like @sieren are doing this from the goodness of their heart . Sorry if I seemed to come across as ungracious or pestering. I truly appreciate AudioKit and team's amazing efforts and their incredible achievements in open source. A million thanks!

I looked at the soundpipe code and dear lord! -- that's some complex stuff! Unfortunately, this is * way way* above my level of expertise and knowledge. I guess, time to start studying DSP, C language, FFTs etc. First step is to understand how this code works!

**Comment by sieren** [comment]
Worth noting that just because I promised I'll take a look at it, does by no means mean I can actually get this job done. I've joined AudioKit because I don't code as part of my dayjob anymore and this is a fun way to stay up to date. But these kinds of optimisations are challenging, given I haven't written DSP Code in years.

**Comment by aure** [comment]
Can't reopen because the code was moved to SoundpipeAudioKit.  Make a new issue there and reference this conversation if you want.

**Comment by quantamrhino** [comment]
Oh thanks - it's done now.
